gpjax/integrators.py

```diff
@@ -0,0 +1,84 @@
+from abc import abstractmethod
+from dataclasses import dataclass
+
+from beartype.typing import (
+    Any,
+    Callable,
+)
+import jax.numpy as jnp
+from jaxtyping import Float
+import numpy as np
+from simple_pytree import Pytree
+
+from gpjax.typing import Array
+
+
+@dataclass
+class AbstractIntegrator(Pytree):
+    @abstractmethod
+    def integrate(
+        self,
+        fun: Callable,
+        y: Float[Array, "N D"],
+        mean: Float[Array, "N D"],
+        sigma2: Float[Array, "N D"],
+        **likelihood_params: Any,
+    ):
+        raise NotImplementedError("self.integrate not implemented")
+
+    def __call__(
+        self,
+        fun: Callable,
+        y: Float[Array, "N D"],
+        mean: Float[Array, "N D"],
+        sigma2: Float[Array, "N D"],
+        *args: Any,
+        **kwargs: Any,
+    ):
+        return self.integrate(fun, y, mean, sigma2, *args, **kwargs)
+
+
+@dataclass
+class GHQuadratureIntegrator(AbstractIntegrator):
+    num_points: int = 20
+
+    def integrate(
+        self,
+        fun: Callable,
+        y: Float[Array, "N D"],
+        mean: Float[Array, "N D"],
+        sigma2: Float[Array, "N D"],
```

Comment:
I would be explicit with `sigma2`, renaming to `variance`? Or explain in a docstring.